terraform/transform_provider.go

```diff
@@ -258,6 +318,94 @@ func (n *graphNodeDisabledProviderFlat) DependentOn() []string {
 	return result
 }
 
+type graphNodeCloseProvider struct {
+	ProviderNameValue string
+}
+
+func (n *graphNodeCloseProvider) Name() string {
+	return fmt.Sprintf("provider.%s (close)", n.ProviderNameValue)
+}
+
+// GraphNodeEvalable impl.
+func (n *graphNodeCloseProvider) EvalTree() EvalNode {
+	return CloseProviderEvalTree(n.ProviderNameValue)
+}
+
+// GraphNodeDependable impl.
+func (n *graphNodeCloseProvider) DependableName() []string {
+	return []string{n.Name()}
+}
+
+func (n *graphNodeCloseProvider) CloseProviderName() string {
+	return n.ProviderNameValue
+}
+
+// GraphNodeDotter impl.
+func (n *graphNodeCloseProvider) DotNode(name string, opts *GraphDotOpts) *dot.Node {
+	return dot.NewNode(name, map[string]string{
+		"label": n.Name(),
+		"shape": "diamond",
+	})
+}
+
+// GraphNodeDotterOrigin impl.
+func (n *graphNodeCloseProvider) DotOrigin() bool {
+	return true
+}
+
+// GraphNodeFlattenable impl.
+func (n *graphNodeCloseProvider) Flatten(p []string) (dag.Vertex, error) {
+	return &graphNodeCloseProviderFlat{
+		graphNodeCloseProvider: n,
+		PathValue:              p,
+	}, nil
+}
+
+// Same as graphNodeCloseProvider, but for flattening
+type graphNodeCloseProviderFlat struct {
+	*graphNodeCloseProvider
+
+	PathValue []string
+}
+
+func (n *graphNodeCloseProviderFlat) Name() string {
+	return fmt.Sprintf(
+		"%s.%s", modulePrefixStr(n.PathValue), n.graphNodeCloseProvider.Name())
+}
+
+func (n *graphNodeCloseProviderFlat) Path() []string {
+	return n.PathValue
+}
+
+func (n *graphNodeCloseProviderFlat) ProviderName() string {
+	return fmt.Sprintf(
+		"%s.%s", modulePrefixStr(n.PathValue),
+		n.graphNodeCloseProvider.CloseProviderName())
+}
+
+// GraphNodeDependable impl.
+func (n *graphNodeCloseProviderFlat) DependableName() []string {
+	return []string{n.Name()}
+}
+
+func (n *graphNodeCloseProviderFlat) DependentOn() []string {
+	var result []string
+
+	// If we're in a module, then depend on our parent's provider
+	if len(n.PathValue) > 1 {
+		prefix := modulePrefixStr(n.PathValue[:len(n.PathValue)-1])
+		if prefix != "" {
+			prefix += "."
+		}
+
+		result = append(result, fmt.Sprintf(
+			"%s%s",
+			prefix, n.graphNodeCloseProvider.Name()))
+	}
+
+	return result
+}
```

Comment:
Wow making a new node type requires a whole lot of noise doesn't it?! :stuck_out_tongue: Probably something we should make a pass on eventually.

/cc @mitchellh